source/common/stats/stats_impl.h

```diff
@@ -283,33 +272,53 @@ class MetricImpl : public virtual Metric {
   const std::vector<Tag> tags_;
 };
 
-/**
- * Implements a StatDataAllocator that uses RawStatData -- capable of deploying
- * in a shared memory block without internal pointers.
- */
-class RawStatDataAllocator : public StatDataAllocator {
+// Partially implements a StatDataAllocator, leaving alloc & free for subclasses.
+// We templatize on StatData rather than defining a virtual base StatData class
+// for performance reasons; stat increment is on the hot path.
+//
+// The two production derivations cover using a fixed block of shared-memory for
+// hot restart stat continuity, and heap allocation for more efficient RAM usage
+// for when hot-restart is not required.
+//
+// Also note that RawStatData needs to live in a shared memory block, and it's
+// possible, but not obvious, that a vptr would be usable across processes. In
+// any case, RawStatData is allocated from a shared-memory block rather than via
+// new, so the usual C++ compiler assistance for setting up vptrs will not be
+// available. This could be resolved with placed new, or another nesting level.
+template <class StatData> class StatDataAllocatorImpl : public StatDataAllocator {
 public:
+  /**
+   * @param stats_options The stat options, which must live longer than the allocator.
+   */
+  StatDataAllocatorImpl() {}
```

Comment:
nit: Is this default constructor needed? Doc comment seems out of date also.

Reply:
oops; not needed; in an earlier rev I needed a 1-arg constructor and removing the arg, I forgot to remove the whole thing.